### PR TITLE
Add bulk delete method to BlobStore interface and implementations

### PR DESCRIPTION
**Pull request by fcofdez**

### CODE REVIEW

**Review by Tim-Brooks — APPROVED** [review]
This looks good to me. But it is not an area of the code that I deal with much so probably needs someone else to take a look over.

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java [hunk 1]
```diff
@@ -181,32 +181,17 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
+                return blobNames.hasNext();
             }
-        }
-        if (ioe != null) {
-            if (suppressedExceptions > 0) {
-                ioe.addSuppressed(new IOException("Failed to delete files, suppressed [" + suppressedExceptions + "] failures"));
+
+            @Override
+            public String next() {
+                return path.resolve(blobNames.next()).toString();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:192]: Could we use `org.elasticsearch.common.collect.Iterators#map` here?

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java [hunk 2]
```diff
@@ -61,6 +63,36 @@ public BlobContainer blobContainer(BlobPath path) {
         return new FsBlobContainer(this, path, f);
     }
 
+    @Override
+    public void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) throws IOException {
+        IOException ioe = null;
+        long suppressedExceptions = 0;
+        while (blobNames.hasNext()) {
+            try {
+                Path resolve = path.resolve(blobNames.next());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java:72]: Hmm this resolves the blob paths again. I think `FsBlobContainer#path` already includes the base path for the repo, so `blobNames.next()` should already be an absolute path, and therefore this should be a no-op.

But then I see that we've got to convert the `String` name to a `Path` somehow, and maybe this is best. Deserves a comment at least.

File: x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalysisSuccessIT.java [hunk 3]
```diff
@@ -240,6 +240,11 @@ private void deleteContainer(AssertingBlobContainer container) {
             }
         }
 
+        @Override
+        public void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) {
+
+        }
```
Comment (DaveCTurner) [inline x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalysisSuccessIT.java:246]: Whitespace nit:

```suggestion
        public void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) {}
```

**Review by DaveCTurner — COMMENTED** [review]
Looks ok to me too - I left a few small comments.

**Comment by fcofdez** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Review by original-brownbear — APPROVED** [review]
LGTM looks safe to me

**Comment by fcofdez** [comment]
Thanks all for the reviews!

**Pull request merged**